File: bot_guidance/src/control_simple.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float64.hpp"
#include <cmath>
#include <geometry_msgs/msg/twist.hpp>
#include <geometry_msgs/msg/pose.hpp>
#include <std_msgs/msg/bool.hpp>
#include <tf2/LinearMath/Matrix3x3.h>
#include <tf2/utils.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>

using namespace std;

double yaw = 0;               // cap a suivre
vector<double> X = {0, 0, 0}; // position robot
vector<double> Y = {0, 0, 0}; // position cible

class Subscriber_publisher : public rclcpp::Node
{
public:
  Subscriber_publisher()
      : Node("simple_controler")
  {
    commande_pub = this->create_publisher<geometry_msgs::msg::Twist>("/demo/cmd_vel", 1000);
    timer_ = this->create_wall_timer(500ms, std::bind(&Subscriber_publisher::timer_callback, this));
    subscriptionBotPose_ = this->create_subscription<geometry_msgs::msg::Pose>("/robot_position", 1000, bind(&Subscriber_publisher::PositionCallBack, this, placeholders::_1));
    subscriptionGoalPose_ = this->create_subscription<geometry_msgs::msg::Pose>("/target", 1000, bind(&Subscriber_publisher::PositionCibleCallBack, this, placeholders::_1));
  }

private:
  void PositionCallBack(const geometry_msgs::msg::Pose::SharedPtr message)
  {
    // position du robot dans le repere du terrain
    X[0] = message->position.x;
    X[1] = message->position.y;
    X[2] = 0.;
    yaw = tf2::getYaw(message->orientation);
    cout << yaw << endl;
  }

  void PositionCibleCallBack(const geometry_msgs::msg::Pose::SharedPtr message)
  {
    // position de la balle de tennis
    Y[0] = message->position.x;
    Y[1] = message->position.y;
    Y[2] = 0.;
  }

  void timer_callback()
  {
    // declaration des variables
    double e;
    double delta_theta;
    double theta_voulu;
    geometry_msgs::msg::Twist msg;
    double k = 8;

    theta_voulu = atan2(Y[1] - X[1], Y[0] - X[0]);

    delta_theta = theta_voulu - yaw;
    e = 2 * atan(tan(delta_theta / 2));
    msg.angular.z = k * e;

    // publication du message
    double dx, dy;
    double k_linear = 1;
    dx = X[0] - Y[0];
    dy = X[1] - Y[1];
    double distance = sqrt(dx * dx + dy * dy);
    msg.linear.x = k_linear * distance;
    commande_pub->publish(msg);
  }

  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::Subscription<geometry_msgs::msg::Pose>::SharedPtr subscriptionBotPose_;
  rclcpp::Subscription<geometry_msgs::msg::Pose>::SharedPtr subscriptionGoalPose_;
  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr commande_pub;
};

int main(int argc, char *argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Subscriber_publisher>());
  rclcpp::shutdown();
  return 0;
}

File: bot_guidance/src/control_simple.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseStamped, Twist, Pose
import numpy as np
import math

# Variable
x_target, y_target = None, None
x, y, yaw = None, None, None


def euler_from_quaternion(quaternion):
    """
    Converts quaternion (w in last place) to euler roll, pitch, yaw
    quaternion = [x, y, z, w]
    Bellow should be replaced when porting for ROS 2 Python tf_conversions is done.
    """
    x = quaternion.x
    y = quaternion.y
    z = quaternion.z
    w = quaternion.w

    sinr_cosp = 2 * (w * x + y * z)
    cosr_cosp = 1 - 2 * (x * x + y * y)
    roll = np.arctan2(sinr_cosp, cosr_cosp)

    sinp = 2 * (w * y - z * x)
    pitch = np.arcsin(sinp)

    siny_cosp = 2 * (w * z + x * y)
    cosy_cosp = 1 - 2 * (y * y + z * z)
    yaw = np.arctan2(siny_cosp, cosy_cosp)

    return roll, pitch, yaw


class controlSimple(Node):
    def __init__(self):
        print(1)
        super().__init__("control_simple")
        print(2)
        self.subscription = self.create_subscription(
            PoseStamped, "robot_position", self.robot_position_callback, 10)
        self.subscription

        self.subscription2 = self.create_subscription(
            Pose, "target", self.target_callback, 10)
        self.subscription2

        self.publisher = self.create_publisher(Twist, '/demo/cmd_vel', 10)
        self.publisher

        timer_period = 0.1  # seconds
        self.timer = self.create_timer(timer_period, self.process)
        self.get_logger().info(self.get_name() + " is launched")

    def process(self):

        msg2 = Twist()
        msg2.linear.x = 2.0
        msg2.angular.z = 2.0
        self.publisher.publish(msg2)

    def robot_position_callback(self, msg):
        global x, y, yaw
        x = msg.pose.position.x
        y = msg.pose.position.y
        quat = msg.pose.orientation
        roll, pitch, yaw = euler_from_quaternion(quat)
        print("robot position:",x,y,yaw)

    def target_callback(self, msg):
        global x_target, y_target
        x_target = msg.position.x
        y_target = msg.position.y
        print("target position:",x_target,y_target)


def main(args=None):
    print(-2)
    rclpy.init(args=args)
    print(-1)
    my_py_node = controlSimple()
    print(0)
    rclpy.spin(my_py_node)

    my_py_node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()
